pkg/agent/agent_linux.go

```diff
@@ -46,3 +187,96 @@ func (i *Initializer) getTunnelPortLocalIP() net.IP {
 func GetTransportIPNetDeviceByName(ifaceName string, ovsBridgeName string) (*net.IPNet, *net.Interface, error) {
 	return util.GetIPNetDeviceByName(ifaceName)
 }
+
+// saveHostRoutes saves routes which are configured on uplink interface before
+// the interface the configured as the uplink of antrea network.
+// The routes will be restored on OVS bridge interface after the IP configuration
+// is moved to the OVS bridge.
+func (i *Initializer) saveHostRoutes() error {
+	routes, err := netlink.RouteList(nil, netlink.FAMILY_V4)
+	if err != nil {
+		return err
+	}
+	for _, route := range routes {
+		if route.LinkIndex != i.nodeConfig.UplinkNetConfig.Index {
+			continue
+		}
+		if route.Gw == nil {
+			continue
+		}
+		// Skip IPv6 routes before we support IPv6 stack.
+		if route.Gw.To4() == nil {
+			continue
+		}
+		klog.Infof("Got host route: %v", route)
+		i.nodeConfig.UplinkNetConfig.Routes = append(i.nodeConfig.UplinkNetConfig.Routes, route)
+	}
+	return nil
+}
+
+// restoreHostRoutes restores the host routes which are lost when moving the IP
+// configuration of uplink interface to the OVS bridge interface during
+// the antrea network initialize stage.
+// The backup routes are restored after the IP configuration change.
+func (i *Initializer) restoreHostRoutes() error {
+	return i.restoreHostRoutesToInterface(i.ovsBridge)
+}
+
+func (i *Initializer) restoreHostRoutesToInterface(ifaceName string) error {
+	iface, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		return nil
+	}
+	for _, routeInterface := range i.nodeConfig.UplinkNetConfig.Routes {
+		route := routeInterface.(netlink.Route)
+		newRoute := route
+		newRoute.LinkIndex = iface.Index
+		if err := netlink.RouteReplace(&newRoute); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RestoreOVSBridge returns immediately on Linux if enableBridge is false.
+// OVS is managed by Antrea in Linux, network config must be restored to uplink before Antrea shutdown.
+func (i *Initializer) RestoreOVSBridge() {
+	if !i.enableBridge {
+		return
+	}
+	klog.Infof("Restore bridge config to uplink...")
+	uplinkNetConfig := i.nodeConfig.UplinkNetConfig
+	uplink := ""
+	if uplinkNetConfig != nil {
+		uplink = uplinkNetConfig.Name
+	}
+
+	brName := i.ovsBridgeClient.GetBridgeName()
+	brLink, err := netlink.LinkByName(brName)
+	if err != nil {
+		klog.Warningf("OVS bridge local port not found: %+v", err)
+	}
+
+	if uplink != "" {
+		if err := util.DeleteOVSPort(brName, uplink); err != nil {
+			klog.Errorf("Remove uplink port %s from %s failed, err: %+v", uplink, brName, err)
+		}
+		if err := util.ConfigureLinkAddresses(uplinkNetConfig.Index, []*net.IPNet{uplinkNetConfig.IP}); err != nil {
+			klog.Errorf("Configure IP to %s failed, err: %+v", uplink, err)
+		}
+	}
+	if brLink != nil {
```

Comment:
Request for clarification: why Is it ok to execute this code block if uplink is ""?